src/Microsoft.OData.Edm/EdmModelVisitor.cs

```diff
@@ -383,11 +596,35 @@ public void VisitSchemaType(IEdmType definition)
             }
         }
 
+        public Task VisitSchemaTypeAsync(IEdmType definition)
+        {
+            switch (definition.TypeKind)
+            {
+                case EdmTypeKind.Complex:
+                    return this.ProcessComplexTypeAsync((IEdmComplexType)definition);
+                case EdmTypeKind.Entity:
+                    return this.ProcessEntityTypeAsync((IEdmEntityType)definition);
+                case EdmTypeKind.Enum:
+                    return this.ProcessEnumTypeAsync((IEdmEnumType)definition);
+                case EdmTypeKind.TypeDefinition:
+                    return this.ProcessTypeDefinitionAsync((IEdmTypeDefinition)definition);
+                case EdmTypeKind.None:
+                    return this.VisitSchemaTypeAsync(definition);
+                default:
+                    throw new InvalidOperationException(Edm.Strings.UnknownEnumVal_TypeKind(definition.TypeKind));

```

Comment:
See my reasoning elsewhere https://github.com/OData/odata.net/pull/3081/files#r1796808438

Reply:
Same here. Resolved by using Contract.Assert and return faulted task:
```cs
...
 Contract.Assert(false, Edm.Strings.UnknownEnumVal_ExpressionKind(expression.ExpressionKind));
 return Task.FromException<InvalidOperationException>(new InvalidOperationException(Edm.Strings.UnknownEnumVal_ExpressionKind(expression.ExpressionKind)));
...
```